ga4gh/backends.py

```diff
@@ -368,99 +311,135 @@ def getMetadata(self):
+    The GA4GH backend. This class provides methods for all of the GA4GH
+    protocol end points.
     """
-    # TODO: create real parent class for VariantSet.
     def __init__(self, dataDir, variantSetClass):
         self._dataDir = dataDir
-        print(dataDir)
-        self._variantSetMap = {}
+        self._variantSetIdMap = {}
         # All directories in datadir are assumed to correspond to VariantSets.
         for variantSetId in os.listdir(self._dataDir):
             relativePath = os.path.join(self._dataDir, variantSetId)
-            # A variant set is all files under a directory,
-            # so we skip non-directory files.
-            if os.path.isfile(relativePath):
-                continue
-            print(variantSetId)
-            self._variantSetMap[variantSetId] = variantSetClass(
-                variantSetId, relativePath)
-        self._variantSetIds = sorted(self._variantSetMap.keys())
+            if os.path.isdir(relativePath):
+                self._variantSetIdMap[variantSetId] = variantSetClass(
+                    variantSetId, relativePath)
+        self._variantSetIds = sorted(self._variantSetIdMap.keys())
 
-    def searchVariants(self, request):
-        assert len(request.variantSetIds) > 0
-        variantSetIndex = 0
-        if request.pageToken is not None:
-            # parse the pageToken and change what will be passed in
-            variantSetIndex, pageToken = request.pageToken.split(':')
-            variantSetIndex = int(variantSetIndex)
-            if pageToken == '':
-                pageToken = None  # None is represented as '' in the pageToken
-            else:
-                pageToken = int(pageToken)
-            request.pageToken = pageToken
-        variantSet = self._variantSetMap[
-            request.variantSetIds[variantSetIndex]]
-        response = variantSet.searchVariants(request)
-        # Add the index of the variant set of the next
-        # page to the nextPageToken
-        if response.nextPageToken is None:
-            if variantSetIndex < len(request.variantSetIds) - 1:
-                # if not, there are no more results, so leave the token as None
-                response.nextPageToken = '{0}:'.format(variantSetIndex + 1)
-                # this token will give results from the beginning of the
-                # next variantSet
-        else:
-            response.nextPageToken = "{0}:{1}".format(variantSetIndex,
-                                                      response.nextPageToken)
-        return response
+    def parsePageToken(self, pageToken, numValues):
+        """
+        Parses the specified pageToken and returns a list of the specified
+        number of values. Page tokens are assumed to consist of a fixed
+        number of integers seperated by colons. If the page token does
+        not conform to this specification, raise a InvalidPageToken
+        exception.
+        """
+        tokens = pageToken.split(":")
```

Comment:
Well, `nextPageToken` is a string in the protocol, so we can safely assume that conforms to this (once we start calling `validate()` on our input request). We've already tested that it's not `None` when we call this method.

We're changing the way things are done slightly here, as pageTokens will not be exactly the same form as before. I think this is a lot simpler though, and should generalise nicely to working with dataSets too. We do have reasonably good test coverage on the paging code, so I'm fairly sure it works properly in the nominal case. We definitely need to write some test code to break it though, and deal with these error cases properly.